.storybook/decorators/PropCombinations/components/PropCombinations.tsx

```diff
@@ -0,0 +1,88 @@
+import { cx } from '@leafygreen-ui/emotion';
+import { GeneratedStoryConfig } from '@leafygreen-ui/lib';
+import { Args, StoryFn } from '@storybook/react';
+import React, { ReactElement } from 'react';
+import { instanceClassName, instanceStyles } from '../PropCombinations.styles';
+import { PropDetailsComponent } from './PropDetails';
+import { shouldExcludePropCombo } from '../utils';
+
+/**
+ * Generates all combinations of each variable
+ */
+export function PropCombinations<T extends React.ComponentType<any>>({
+  component,
+  variables,
+  args,
+  exclude,
+  decorator = (SFn: StoryFn) => <SFn />,
+}: {
+  component: T;
+  variables: Array<[string, Array<any> | undefined]>;
+  args: Args;
+  exclude: GeneratedStoryConfig<T>['excludeCombinations'];
+  decorator: GeneratedStoryConfig<T>['decorator'];
+}): ReactElement<any> {
+  let comboCount = 0;
+
+  const AllCombinations = RecursiveCombinations({}, [...variables]);
+
+  console.log(
+    `Rendering ${comboCount} prop combinations for component: ${component.displayName}`,
+  );
+
+  return AllCombinations;
+
+  /**
+   * Recursively loop through all prop combinations defined in `variables` and render them
+   */
+  function RecursiveCombinations(
```

Comment:
Nit – I'm not in love with unreachable statements in written code, even if hoisting fixes it under the hood.

We also seem to be passing some values as arguments to this function (like variables), and leveraging others that are in scope where this function is defined (like `args`). Can we take one approach or the other here? Bonus points if we move this function outside of the `PropCombinations` function altogether and make it more testable on its own.

Reply:
For recursive functions I think scoping it within a closure is the best way to handle variables that need to be accessible regardless of recursion level. The alternative is passing the same prop down each iteration, which afaik is not recommended.

The other functions defined below could be defined externally and take a pass through prop

Reply:
As for were the function is defined, I see 2 camps, each answering a different question when reading the outer function.

When I read the outer function, I ask:

Q1:  "what is this function doing"?
Q2: "where is this dependency defined"

As written, this function optimizes for answering Q1 while leaving Q2 harder to answer.
Moving the recursive function definition to the top, and return statement to the bottom reverses this and optimizes for Q2, while making Q1 harder to answer.

IMO Q1 is the more important question